keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,724 @@
+mode will be assumed to preserve existing behavior.
+
+### Changes to external-provisioner
+
+This design leverages the access rights of a user on `VolumeSnapshotContents` to
+determine whether the volume mode can be modified when a `PVC` is being created
+with a `VolumeSnapshot` as the source.
+The volume mode can be altered if the requesting user has `Update` and `Patch` rights
+on `VolumeSnapshotContents` (which is a cluster scoped resource).
+The control flow for creating a `PVC` from a `VolumeSnapshot` will look like below:
+
+1. A user attempts to create a `PVC` from a `VolumeSnapshot` by specifying the
+`Spec.DataSource` parameter of the `PVC` YAML.
+2. `external-provisioner` receives a callback to dynamically create the volume.
+As part of the preprocessing steps, it will:
+   1. Get the `Spec.SourceVolumeMode` of the `VolumeSnapshotContent`.
+      1. If `Spec.SourceVolumeMode` doesn't exist or is nil, then continue with
+      volume provisioning to preserve existing behavior.
+   2. Get the `Spec.VolumeMode` of the `PVC` being created.
+   If they do not match:
+      1. Get all annotations on the `VolumeSnapshotContent` and verify if 
+      `snapshot.storage.kubernetes.io/allowVolumeModeChange: true` exists.
+      If it does not exist, block volume provisioning by returning an error.
+4. In all other cases, let volume provisioning continue.
+
+NOTE: `external-provisioner` maintains a reference to `PVC` and `VolumeSnapshotContent`
+during volume creation. This proposal leverages those references to make additional
+decisions.
+
+### Test Plan
+
+E2E tests will be added for this design, that attempt to restore a volume with
+and without requisite privileges. 
+
+#### Unit tests
+
+- With feature gate disabled:
+  - attempt to convert volume mode when creating a `PVC`
+  from a `VolumeSnapshot`.
+- With feature gate enabled, attempt to convert volume mode when creating a `PVC`
+from a `VolumeSnapshot`:
+  - With `Spec.SourceVolumeMode` populated and `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+  annotation present.
+  - With `Spec.SourceVolumeMode` populated but no `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+  annotation.
+  - With `Spec.SourceVolumeMode` set to `nil`.
+
+#### E2E tests
+
+- Attempt to convert volume mode when creating a `PVC`from a `VolumeSnapshot`:
+  - With `Spec.SourceVolumeMode` populated and `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+    annotation present.
+  - With `Spec.SourceVolumeMode` populated but no `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+    annotation.
+  - With `Spec.SourceVolumeMode` set to `nil`.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind an out-of-tree feature flag.
+- Feedback from users.
+- Implementation of unit and e2e tests.
+
+#### Alpha -> Beta
+
+- One release with positive feedback from users. 
+
+#### Beta -> GA
+
+- Deployed in production in use by backup vendors. 
```

Comment:
backup vendors -> backup software
There should be users running backup software which uses this feature at this point.

Reply:
Changed